src/go/rdctl/pkg/config/config.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright © 2022 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package config handles all the config-related parts of rdctl
+
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/spf13/cobra"
+)
+
+// The CLIConfig struct is used to store the json data read from the config file.
+type CLIConfig struct {
+	User     string
+	Password string
+	Port     int
+}
+
+type ConnectionInfo struct {
+	User     string
+	Password string
+	Host     string
+	Port     string
+}
+
+var (
+	connectionSettings ConnectionInfo
+
+	configDir         string
+	configPath        string
+	defaultConfigPath string
+)
+
+// DefineGlobalFlags sets up the global flags, available for all sub-commands
+func DefineGlobalFlags(rootCmd *cobra.Command) {
+	var err error
+
+	configDir, err = os.UserConfigDir()
+	if err != nil {
+		log.Fatal("Can't get config-dir: ", err)
+	}
+	defaultConfigPath = filepath.Join(configDir, "rancher-desktop", "rd-engine.json")
+	rootCmd.PersistentFlags().StringVar(&configPath, "config-path", "", fmt.Sprintf("config file (default %s)", defaultConfigPath))
+	rootCmd.PersistentFlags().StringVar(&connectionSettings.User, "user", "", "overrides the user setting in the config file")
+	rootCmd.PersistentFlags().StringVar(&connectionSettings.Host, "host", "", "default is localhost; most useful for WSL")
+	rootCmd.PersistentFlags().StringVar(&connectionSettings.Port, "port", "", "overrides the port setting in the config file")
+	rootCmd.PersistentFlags().StringVar(&connectionSettings.Password, "password", "", "overrides the password setting in the config file")
+}
+
+// GetConnectionInfo returns the connection info if it has it, and an error message explaining why
+// it isn't available if it doesn't have it.
+// So if the user runs an `rdctl` command after a factory reset, there is no config file (in the default location),
+// but it might not be necessary. So only use the error message for the missing file if it is actually needed.
+func GetConnectionInfo() (*ConnectionInfo, error) {
+	err, isImmediateError := initConfig()
```

Comment:
We should do something to prevent `initConfig` from being called multiple times; it's harmless but confusing for something called `init…`. Hence the [`sync.Once`](https://pkg.go.dev/sync#Once) suggestion earlier.

Reply:
In a larger system I could see a case being made for this, but the code is small and straightforward enough -- `GetConnectionInfo` is invoked on two different code paths (one for GET requests, one for POST requests), and nothing else calls it.

I had a look at `sync.Once` and using that would introduce other complications because it's geared more to `void` functions, and we'd need a wrapper.

This seems like more of a consideration in an environment with more hostile actors, but this is a standalone client.

And I don't see any problems if in fact the function is called subsequent times. Most of the time it updates parameters that aren't set yet, and they should all be set after the first call.

As a compromise, I renamed the function to `finishConnectionSettings`